### PR TITLE
[FLINK-17313]fix type validation error when sink table ddl exists column with precision of decimal/varchar

### PR DESCRIPTION
**Pull request by zjuwangg**
## What is the purpose of the change

* Now TypeMappingUtils#checkPhysicalLogicalTypeCompatible method doesn't consider the different physical and logical type validation logic of source and sink: logical type should be able to cover the physical type in source, but physical type should be able to cover the logic type in sink vice verse. Besides, the decimal type should be taken more carefully, when target type is Legacy(Decimal), it should be able to accept any precision decimal type.*
* This pr aims to solve the above problem.*

## Brief change log

  - 9b24a40 correct the validation logc

## Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no )

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java [hunk 1]
```diff
@@ -158,19 +158,33 @@ public static void checkPhysicalLogicalTypeCompatible(
...
+					physicalFieldType,
+					logicalFieldType,
+					(cause) -> new ValidationException(
+								String.format("Type %s of table field '%s' does not match with " +
+									"the physical type %s of the '%s' field of the %s.",
+										logicalFieldType,
+										logicalFieldName,
+										physicalFieldType,
+										physicalFieldName,
+										"TableSource return type"),
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java:172]: Lets just inline the last argument.

Comment (zjuwangg) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java:172]: ok

**Comment by zjuwangg** [comment]
Thanks for your detailed review @dawidwys 
I'll address your comments and add tests soon.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java [hunk 2]
```diff
@@ -169,8 +169,7 @@ public static void checkPhysicalLogicalTypeCompatible(
 										logicalFieldName,
 										physicalFieldType,
 										physicalFieldName,
-										"TableSource return type"),
-								cause));
+										"TableSource return type"), cause));
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java:172]: What I meant is that we can inline the `TableSource return type` into the string. There is no reason for having it as an argument to `String.format()`.:
```
(cause) -> new ValidationException(
						String.format("Type %s of table field '%s' does not match with " +
							"the physical type %s of the '%s' field of the "TableSource return type".",
								logicalFieldType,
								logicalFieldName,
								physicalFieldType,
								physicalFieldName),
						cause));
```

Comment (zjuwangg) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TypeMappingUtils.java:172]: Oh, sry, you are right.

**Comment by dawidwys** [comment]
I will review it today.

**Comment by dawidwys** [comment]
Thank you for the update. Merged.

**Comment by zjuwangg** [comment]
> Thank you for the update. Merged.

Thanks your review and merge 🌹 @dawidwys

**Pull request merged**